src/common/src/util/value_encoding/mod.rs

```diff
@@ -31,6 +35,190 @@ use error::ValueEncodingError;
+}
+
+impl RowEncoding {
+    pub fn new() -> Self {
+        RowEncoding {
+            flag: 0b_1000_0000,
+            non_null_datum_nums: Width::Mid(0),
+            non_null_column_ids: vec![],
+            offsets: vec![],
+            buf: vec![],
+        }
+    }
+
+    pub fn set_width(&mut self, non_null_column_nums: usize) {
+        self.non_null_datum_nums = match non_null_column_nums {
+            n if n <= u8::MAX as usize => {
+                self.flag |= 0b0100;
+                Width::Mid(n as u8)
+            }
+            n if n <= u16::MAX as usize => {
+                self.flag |= 0b1000;
+                Width::Large(n as u16)
+            }
+            n if n <= u32::MAX as usize => {
+                self.flag |= 0b1100;
+                Width::Extra(n as u32)
+            }
+            _ => unreachable!("the number of columns exceeds u32"),
+        }
+    }
+
+    pub fn set_big(&mut self, maybe_offset: Vec<usize>, max_offset: usize) {
+        self.offsets = vec![];
+        match max_offset {
+            n if n <= u8::MAX as usize => {
+                self.flag |= 0b01;
+                maybe_offset
+                    .into_iter()
+                    .for_each(|m| self.offsets.put_u8(m as u8));
+            }
+            n if n <= u16::MAX as usize => {
+                self.flag |= 0b10;
+                maybe_offset
+                    .into_iter()
+                    .for_each(|m| self.offsets.put_u16_le(m as u16));
+            }
+            n if n <= u32::MAX as usize => {
+                self.flag |= 0b11;
+                maybe_offset
+                    .into_iter()
+                    .for_each(|m| self.offsets.put_u32_le(m as u32));
+            }
+            _ => unreachable!("encoding length exceeds u32"),
+        }
+    }
+
+    pub fn encode(
+        &mut self,
+        datum_refs: impl Iterator<Item = impl ToDatumRef>,
+        column_ids: Vec<ColumnId>,
+    ) {
+        assert!(
+            self.buf.is_empty() && self.non_null_column_ids.is_empty(),
+            "should not encode one RowEncoding object multiple times."
+        );
+        let mut maybe_offset = vec![];
+        for (datum, &id) in datum_refs.zip_eq(column_ids.iter()) {
+            if let Some(v) = datum.to_datum_ref() {
+                maybe_offset.push(self.buf.len());
+                self.non_null_column_ids.push(id);
+                serialize_scalar(v, &mut self.buf);
+            }
+        }
+        let max_offset = *maybe_offset
+            .last()
+            .expect("should encode at least one column");
+        self.set_width(self.non_null_column_ids.len());
+        self.set_big(maybe_offset, max_offset);
+    }
+
+    pub fn serialize(&self) -> Vec<u8> {
+        let mut row_bytes = vec![];
+        row_bytes.put_u8(self.flag);
+        serialize_width(self.non_null_datum_nums, &mut row_bytes);
+        for id in &self.non_null_column_ids {
+            row_bytes.put_i32_le(id.get_id());
```

Comment:
I think we should choose between u8, u16, and u32 for column ids, but it seems we're using signed integers for column IDs now? 😄 May discuss it later.